common/clock/ratelimiter.go

```diff
@@ -0,0 +1,418 @@
+	if tryNow.After(r.latestNow) {
+		// this should never occur!
+		//
+		// reservations and their reserved-time are gathered behind a lock, and time is monotonic.
+		// that means that a reserved-time should ALWAYS be the same or older than any other time
+		// the ratelimiter sees -> its time would have to go backwards.
+		//
+		// that said, if this DOES happen, it must still maintain the newest time.
+		// this should only occur in tests that move time backwards, unless there's
+		// an implementation bug in the ratelimiter or Go's time.
+		r.latestNow = tryNow
+	}
+	return r.latestNow, r.mut.Unlock
+}
+
+func (r *ratelimiter) Allow() bool {
+	now, unlock := r.lockNow()
+	defer unlock()
+	return r.limiter.AllowN(now, 1)
+}
+
+func (r *ratelimiter) Reserve() Reservation {
+	now, unlock := r.lockNow()
+	defer unlock()
+	res := r.limiter.ReserveN(now, 1)
+	if res.OK() && res.DelayFrom(now) == 0 {
+		// usable token, return a cancel-able handler in case it's not used
+		return &allowedReservation{
+			res:        res,
+			reservedAt: now,
+			limiter:    r,
+		}
+	}
+	// unusable, immediately cancel it to get rid of the future-reservation
+	// since we don't allow waiting for it.
+	//
+	// this restores MANY more tokens when heavily contended, as time skews
+	// less before canceling, so it gets MUCH closer to the intended limit.
+	res.CancelAt(now)
+	return deniedReservation{}
+}
+
+func (r *ratelimiter) Burst() int {
+	// rate.Limiter.Burst does not currently advance time,
+	// so this does not need to hold the ratelimiter lock.
+	// if this changes, use lockNow.
+	return r.limiter.Burst()
+}
+
+func (r *ratelimiter) Limit() rate.Limit {
+	// rate.Limiter.Limit does not currently advance time,
+	// so this does not need to hold the ratelimiter lock.
+	// if this changes, use lockNow.
+	return r.limiter.Limit()
+}
+
+func (r *ratelimiter) SetBurst(newBurst int) {
+	now, unlock := r.lockNow()
+	defer unlock()
+	r.limiter.SetBurstAt(now, newBurst)
+}
+
+func (r *ratelimiter) SetLimit(newLimit rate.Limit) {
+	now, unlock := r.lockNow()
+	defer unlock()
+	r.limiter.SetLimitAt(now, newLimit)
+}
+
+func (r *ratelimiter) Tokens() float64 {
+	now, unlock := r.lockNow()
+	defer unlock()
+	return r.limiter.TokensAt(now)
+}
+
+func (r *ratelimiter) Wait(ctx context.Context) (err error) {
+	now, unlock := r.lockNow()
+	var once sync.Once
+	defer once.Do(unlock) // unlock if panic or returned early with no err
+
+	res := r.limiter.ReserveN(now, 1)
+	defer func() {
+		if err != nil {
+			// err return means "not allowed", so cancel the reservation.
+			//
+			// note that this makes a separate call to get the current time:
+			// this is intentional, as time may have passed while waiting.
+			//
+			// if the cancellation happened before the time-to-act (the delay),
+			// the reservation will still be successfully rolled back.
+
+			// ensure we are unlocked, which will not have happened if an err
+			// was returned immediately.
+			once.Do(unlock)
```

Comment:
slightly confused about the number of unlocks here. My recollection is that defers are LIFO and you're manually unlocking here. Won't this be called a second time on 327?

Reply:
no, that's what you're using once for :facepalm: ignore me

Reply:
Yea - it just felt (a lot) easier to `once` it and not worry about handling specific branches.  Much harder to screw up.

I can figure out exactly what that'd take if we wanted a different route / see the comparison, and it doesn't actually have to be a *sync*-once to work so it can be optimized a bit.  I just fought with it briefly and then realized "wait this approach is obviously correct, I'll just do that" and never looked back.